File: leetcode-js/361-bomb-enemy.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var maxKilledEnemies = function(grid) {
    let result = 0;
    let rowHits = 0;
    const colHits = Array(grid[0].length).fill(0);

    for (let i = 0; i < grid.length; i += 1) {
        for (let j = 0; j < grid[0].length; j += 1) {
            if (!j || grid[i][j-1] === 'W') {
                rowHits = 0;
                for (let k = j; k < grid[0].length && grid[i][k] !== 'W'; k += 1) {
                    rowHits += grid[i][k] === 'E' ? 1 : 0;
                }
            }
            if (!i || grid[i-1][j] === 'W') {
                colHits[j] = 0;
                for (let k = i; k < grid.length && grid[k][j] !== 'W'; k += 1) {
                    colHits[j] += grid[k][j] === 'E' ? 1 : 0;
                }
            }
            if (grid[i][j] === '0') {
                result = Math.max(result, rowHits + colHits[j]);
            }
        }
    }

    return result;
};

File: leetcode-java/361-bomb-enemy.java
public class Solution {
    public int maxKilledEnemies(char[][] grid) {
        if (grid.length == 0) {
            return 0;
        }

        int m = grid.length;
        int n = grid[0].length;
        int rowCount = 0;
        int[] colCounts = new int[n];

        int max = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (j == 0 || grid[i][j - 1] == 'W') {
                    rowCount = 0;
                    for (int k = j; k < n && grid[i][k] != 'W'; k++) {
                        if (grid[i][k] == 'E') {
                            rowCount++;
                        }
                    }
                }

                if (i == 0 || grid[i - 1][j] == 'W') {
                    colCounts[j] = 0;
                    for (int k = i; k < m && grid[k][j] != 'W'; k++) {
                        if (grid[k][j] == 'E') {
                            colCounts[j]++;
                        }
                    }
                }

                if (grid[i][j] == '0') {
                    max = Math.max(max, rowCount + colCounts[j]);
                }
            }
        }

        return max;
    }
}
